### PR TITLE
DIV-6705  Respondent Organisation policy field exists

### PR DESCRIPTION
**Pull request by jacmurphy**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/DIV-6705

### Change description ###
Permissions are not finalised yet. Waiting on Jeremy to decide.

UserRole | CRU
-- | --
caseworker-caa | CRU
[RESPSOLICITOR] | R
caseworker-divorce-solicitor | CRU
caseworker-divorce-courtadmin_beta | R
caseworker-divorce-superuser | R
caseworker-divorce-courtadmin | R
caseworker-divorce-courtadmin-la | R
citizen | CRU

### CODE REVIEW

**Review by yarpo — COMMENTED** [review]
Please add jira id to PR title and link to ticket in description.

File: test/unit/definitions/divorce/AuthorisationCaseField.test.js [hunk 1]
```diff
@@ -34,7 +35,8 @@ describe('AuthorisationCaseField', () => {
         'CaseField-deemed-and-dispensed-nonprod',
         'CaseField-general-email-nonprod',
         'CaseField-general-referral-nonprod',
-        'CaseField-share-a-case-nonprod'
+        'CaseField-share-a-case-nonprod',
```
Comment (yarpo) [inline test/unit/definitions/divorce/AuthorisationCaseField.test.js:38]: alphabetical order, please :)

File: test/unit/definitions/divorce/AuthorisationComplexCaseType.test.js [hunk 2]
```diff
@@ -10,9 +10,15 @@ describe('AuthorisationComplexType', () => {
     let allFieldsForNonProd = [];
 
     before(() => {
-      nonProdAuthorisationComplexType = getAuthorisationComplexTypeDefinitions(['AuthorisationComplexType-share-a-case-nonprod']);
+      nonProdAuthorisationComplexType = getAuthorisationComplexTypeDefinitions([
+        'AuthorisationComplexType-share-a-case-nonprod',
+        'AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod'
```
Comment (yarpo) [inline test/unit/definitions/divorce/AuthorisationComplexCaseType.test.js:15]: alphabetical order, please :)

File: test/unit/definitions/divorce/AuthorisationComplexCaseType.test.js [hunk 3]
```diff
@@ -10,9 +10,15 @@ describe('AuthorisationComplexType', () => {
...
+      ]);
 
-      allFieldsForNonProd = getCaseFieldDefinitions(['CaseField-share-a-case-nonprod']);
+      allFieldsForNonProd = getCaseFieldDefinitions([
+        'CaseField-share-a-case-nonprod',
```
Comment (yarpo) [inline test/unit/definitions/divorce/AuthorisationComplexCaseType.test.js:19]: alphabetical order, please :)

**Review by yarpo — APPROVED** [review]

File: test/unit/definitions/divorce/AuthorisationComplexCaseType.test.js (hunk 3 above)
Comment (jacmurphy) [reply test/unit/definitions/divorce/AuthorisationComplexCaseType.test.js]: Done

File: test/unit/definitions/divorce/AuthorisationComplexCaseType.test.js (hunk 2 above)
Comment (jacmurphy) [reply test/unit/definitions/divorce/AuthorisationComplexCaseType.test.js]: Done

File: test/unit/definitions/divorce/AuthorisationCaseField.test.js (hunk 1 above)
Comment (jacmurphy) [reply test/unit/definitions/divorce/AuthorisationCaseField.test.js]: Done

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
+[
+  {
+    "LiveFrom": "26/01/2021",
+    "CaseTypeID": "DIVORCE",
+    "CaseFieldID": "RespondentOrganisationPolicy",
+    "UserRole": "caseworker-caa",
```
Comment (jvictorcf) [inline definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json:6]: what is this role?

File: definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json [hunk 5]
```diff
@@ -0,0 +1,34 @@
+[
+  {
+    "LiveFrom": "26/01/2021",
+    "CaseTypeID": "DIVORCE",
+    "CaseFieldID": "RespondentOrganisationPolicy",
+    "ListElementCode": "OrgPolicyCaseAssignedRole",
+    "UserRole": "[RESPSOLICITOR]",
```
Comment (Dizrupto) [inline definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json:7]: it seems only the "caseworker-caa" needs to have permissions for fields inside the complex type :)
So you would need to replace [RESPSOLICITOR] with caseworker-caa
(see https://github.com/hmcts/finrem-ccd-definitions/blob/230139d4c0fe134cdf216821ed5e7cae054d0d9d/definitions/contested/json/AuthorisationComplexType/AuthorisationComplexType-respSol-nonprod.json#L6)

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
+[
```
Comment (jvictorcf) [inline definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json:1]: @jacmurphy can you please follow the feature-toggle file name that's already used?
I believe that's "resp-journey-roles-and-permissions".
https://github.com/hmcts/div-ccd-definitions/blob/master/definitions/divorce/json/CaseTypeTab/CaseTypeTab-resp-journey-roles-and-permissions-nonprod.json

File: definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    "CaseFieldID": "RespondentOrganisationPolicy",
+    "ListElementCode": "Organisation",
+    "UserRole": "[RESPSOLICITOR]",
+    "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "26/01/2021",
+    "CaseTypeID": "DIVORCE",
+    "CaseFieldID": "RespondentOrganisationPolicy",
+    "ListElementCode": "Organisation.OrganisationID",
```
Comment (jvictorcf) [inline definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json:30]: is this really going to work?
I thought you'd have to create another complex type for this...

Comment (jvictorcf) [reply definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json:30]: Sorry, do we authorisation for complex types? That's new, right?

File: test/unit/definitions/divorce/AuthorisationCaseField.test.js [hunk 8]
```diff
@@ -16,24 +16,26 @@ describe('AuthorisationCaseField', () => {
...
-        'AuthorisationCaseField-alternative-service-nonprod',
         'AuthorisationCaseField-alt-service-process-server-nonprod',
+        'AuthorisationCaseField-alternative-service-nonprod',
+        'AuthorisationCaseField-amend-court-orders-nonprod',
         'AuthorisationCaseField-deemed-and-dispensed-nonprod',
         'AuthorisationCaseField-general-email-nonprod',
         'AuthorisationCaseField-general-referral-nonprod',
-        'AuthorisationCaseField-share-a-case-nonprod',
-        'AuthorisationCaseField-nonprod'
+        'AuthorisationCaseField-nonprod',
```
Comment (jvictorcf) [inline test/unit/definitions/divorce/AuthorisationCaseField.test.js:25]: Personal taste: Can you leave "non-prod" in the bottom of the list (as that's the one that won't be removed in releases)?

Comment (jvictorcf) [reply test/unit/definitions/divorce/AuthorisationCaseField.test.js:25]: or on top. just below the core file.

**Review by jvictorcf — COMMENTED** [review]
A couple of questions for you.
Also, this ticket doesn't tell me much about what you're trying to achieve.
How will this be tested?

File: test/unit/definitions/divorce/AuthorisationCaseField.test.js (hunk 8 above)
Comment (jacmurphy) [reply test/unit/definitions/divorce/AuthorisationCaseField.test.js:25]: Im a fan of that too, however @yarpo wanted it to be alphabetised. But im happy to move forwards with it alphabetised with the exception of keeping default non prod at the top. How does that sound

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json (hunk 6 above)
Comment (jacmurphy) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json:1]: will do

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json (hunk 4 above)
Comment (Dizrupto) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json:6]: We're still not 100% sure but we do know they need permissions for all the organisation policy fields. @madjava might know a bit more.

File: definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json (hunk 7 above)
Comment (Dizrupto) [reply definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json:30]: Yep apparently those are authorisations that can be applied to each different field inside a complex type.

**Review by madjava — APPROVED** [review]

**Review by Dizrupto — APPROVED** [review]

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json (hunk 4 above)
Comment (jacmurphy) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json]: Taken from - https://tools.hmcts.net/confluence/display/RCCD/Access+Control+Scope+of+Delivery?focusedCommentId=1471101476#comment-1471101476

This is an IdAM role.
_As part of the functionality to implement Assign Access to a Case, the IdAM role caseworker-caa was introduced to be applied to a system role only with specific code that applies for this role.  The reason for this is that the system user provides mediated access to Case Access Administrators through privileged access.  This idam-role specific code will be maintained in CCD as the Access Control solution has no real bearing on its behaviour._

Comment (jacmurphy) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-roles-and-permissions-resp-journey-nonprod.json]: @jvictorcf ^

File: definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json (hunk 5 above)
Comment (jacmurphy) [reply definitions/divorce/json/AuthorisationComplexType/AuthorisationComplexType-roles-and-permissions-resp-journey-nonprod.json]: ive updated this area to better reflect the Updated roles, in this commit - a0d346f7391292391a83f5bd635638ae9538f753

**Pull request merged**